index.d.ts

```diff
@@ -0,0 +1,82 @@
+import { Action, StoreEnhancer, Store } from "redux";
+
+export interface StoreCreator {
+  <S, T>(
+    reducer: EnhancedReducer<S>,
+    preloadedState: S,
+    enhancer: StoreEnhancer<S>
+  ): Store<S>;
+}
+
+export interface EnhancedReducer<S> {
+  <A extends Action>(state: S, action: A): S | [S, CmdType<A, Function>];
+}
+
+declare interface LiftedEnhancedReducer<S> {
+  <A extends Action>(state: S, action: A): [S, CmdType<A, Function>];
+}
+
+export interface NoneCmd {
+  type: "NONE";
+}
+
+export interface ActionCmd<A extends Action> {
```

Comment:
what's the reason for making all of the Cmds generics? I get the reducer because we want to make sure the function always returns the right type. But won't you just say something like

```js
let cmd: ActionCmd = Cmd.action(someAction)
return loop(newState, cmd);
```

in the reducer?

Reply:
Good question. I'm starting to think that the `F` generic for the `CmdType` isn't necessary. 

The thinking behind making Action a generic is that in the reducer file you can define a union type of Actions that your reducer expects to handle and then you get an additional layer of typesafety ensuring that your Cmd.action() will be receiving actions that the reducer knows how to handle.

i'd recommend taking a look at the [sample project here](https://github.com/willisplummer/redux-loop-ts/blob/master/src/index.tsx#L8-L19).

for example:

```
type Action = { type: 'INCREMENT', amount: number } | { type: 'INCREMENT_AGAIN' }

const reducer: EnhancedReducer<
  number,
  number | [number, CmdType<Action>]
> = (state = 0, action: Action) => {
  switch (action.type) {
    case 'INCREMENT':
      return loop(state + amount, Cmd.action({type: 'INCREMENT_AGAIN'})
    case 'INCREMENT_AGAIN':
      return state + 1;
    default:
      return state;
  }
}
```

Without the generic if you accidentally wrote a typo like `return loop(state + amount, Cmd.action({ type: 'INCREMENT_AGAN'})` that would be a valid `Cmd.action` because it is a redux `Action` (an object with a `type` property) but your reducer would just fall through to the default case and nothing would happen.  Similarly, if you tried to call `Cmd.action({ type: 'INCREMENT', amount: 'not a number' })` typescript wouldn't be able to catch that for you.

Reply:
but all reducers get called with all actions anyway. So the reducer can't limit what actions it receives. it just needs to know to use the default handler for anything it doesn't recognize.

Reply:
yeah - It's that the types add a guard against what actions the commands can dispatch